reactor-test/src/main/java/reactor/test/subscriber/TestSubscriber.java

```diff
@@ -0,0 +1,417 @@
+	public void onComplete() {
+		Signal<T> sig = Signal.complete();
+		if (terminalSignal.compareAndSet(null, sig)) {
+			notifyDone();
+		}
+		else {
+			this.protocolErrors.add(sig);
+		}
+	}
+
+	@Override
+	public void onError(Throwable t) {
+		Signal<T> sig = Signal.error(t);
+		if (terminalSignal.compareAndSet(null, sig)) {
+			notifyDone();
+		}
+		else {
+			this.protocolErrors.add(sig);
+		}
+	}
+
+	@Nullable
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.TERMINATED) return terminalSignal.get() != null;
+		if (key == Attr.CANCELLED) return cancelled.get();
+		if (key == Attr.ERROR) {
+			Signal<T> sig = terminalSignal.get();
+			if (sig == null || sig.getThrowable() == null) {
+				return null;
+			}
+			return sig.getThrowable();
+		}
+		if (key == Attr.PARENT) return s;
+		if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+
+		return null;
+	}
+
+	// == public subscription-like methods
+
+	public void cancel() {
+		if (cancelled.compareAndSet(false, true)) {
+			if (this.s != null) {
+				this.s.cancel();
+			}
+			notifyDone();
+		}
+	}
+
+	public void request(long n) {
+		if (this.s == null) {
+			throw new IllegalStateException("Request can only happen once a Subscription has been established." +
+					"Have you subscribed the TestSubscriber?");
+		}
+		if (Operators.validate(n)) {
+			if (this.fusionMode == Fuseable.SYNC) {
+				internalCancel();
+				throw new IllegalStateException("Request is short circuited in SYNC fusion mode, and should not be explicitly used");
+			}
+			this.s.request(n);
+		}
+	}
+
+	// == public accessors
+
+	public boolean isTerminated() {
+		return terminalSignal.get() != null;
+	}
+
+	public boolean isCancelled() {
+		return cancelled.get();
+	}
+
+	public boolean isTerminatedOrCancelled() {
+		return terminalSignal.get() != null || cancelled.get();
+	}
+
+	@Nullable
+	public Signal<T> getTerminalSignal() {
+		return this.terminalSignal.get();
+	}
+
+	public Signal<T> checkTerminalSignal() {
+		Signal<T> sig = terminalSignal.get();
+		if (sig == null || (!sig.isOnError() && !sig.isOnComplete())) {
+			cancel();
+			throw new AssertionError("Expected subscriber to be terminated, but it has not been terminated yet.");
+		}
+		return sig;
+	}
+
+	public Throwable checkTerminalError() {
+		Signal<T> sig = terminalSignal.get();
+		if (sig == null) {
+			cancel();
+			throw new AssertionError("Expected subscriber to have errored, but it has not been terminated yet.");
+		}
+		if (sig.isOnComplete()) {
+			throw new AssertionError("Expected subscriber to have errored, but it has completed instead.");
+		}
+		Throwable terminal = sig.getThrowable();
+		if (terminal == null) {
+			cancel();
+			throw new AssertionError("Expected subscriber to have errored, got unexpected terminal signal <" + sig + ">.");
+		}
+		return terminal;
+	}
+
+	public List<T> getReceivedOnNext() {
+		return new ArrayList<>(this.receivedOnNext);
+	}
+
+	public List<T> getOnNextSublistReceivedAfterCancel() {
```

Comment:
`getReceivedOnNextAfterCancellation` aligning with `getReceivedOnNext`?

Reply:
this can probably be conveyed through the javadoc, but one thing I want to point out here is that this is (conceptually) a sublist of the `getReceivedOnNext()` list. ie. the later always contains the former. this method is offered for convenience in case one wants to check how the source behaves wrt cancellation, namely: can it produce "extraneous" elements after cancellation (technically allowed by the RS spec).